testing/multi-env-test-engine/src/main/java/org/projectnessie/junit/engine/ThreadPerTestClassExecutionExecutorService.java

```diff
@@ -0,0 +1,117 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.junit.engine;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+import java.lang.reflect.Field;
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutorService;
+
+/**
+ * Implements a JUnit test executor that provides thread-per-test-class behavior.
+ *
+ * <p>"Thread-per-test-class behavior" is needed to prevent the class/class-loader leak via {@link
+ * ThreadLocal}s as described in <a
+ * href="https://github.com/projectnessie/nessie/issues/9441">#9441</a>.
+ */
+public class ThreadPerTestClassExecutionExecutorService implements HierarchicalTestExecutorService {
+
+  private static final Class<?> CLASS_NODE_TEST_TASK;
+  private static final Field FIELD_TEST_DESCRIPTOR;
+
+  static {
+    try {
+      CLASS_NODE_TEST_TASK =
+          Class.forName("org.junit.platform.engine.support.hierarchical.NodeTestTask");
+      FIELD_TEST_DESCRIPTOR = CLASS_NODE_TEST_TASK.getDeclaredField("testDescriptor");
+      FIELD_TEST_DESCRIPTOR.setAccessible(true);
+    } catch (Exception e) {
+      throw new RuntimeException(
+          "ThreadPerExecutionExecutorService is probably not compatible with the current JUnit version",
+          e);
+    }
+  }
+
+  protected TestDescriptor getTestDescriptor(TestTask testTask) {
+    if (!CLASS_NODE_TEST_TASK.isAssignableFrom(testTask.getClass())) {
+      throw new IllegalArgumentException(
+          testTask.getClass().getName() + " is not of type " + CLASS_NODE_TEST_TASK.getName());
+    }
+    try {
+      return (TestDescriptor) FIELD_TEST_DESCRIPTOR.get(testTask);
+    } catch (IllegalAccessException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public ThreadPerTestClassExecutionExecutorService() {}
+
+  @Override
+  public Future<Void> submit(TestTask testTask) {
+    executeTask(testTask);
+    return completedFuture(null);
+  }
+
+  @Override
+  public void invokeAll(List<? extends TestTask> tasks) {
+    tasks.forEach(this::executeTask);
+  }
+
+  protected void executeTask(TestTask testTask) {
+    TestDescriptor testDescriptor = getTestDescriptor(testTask);
+    UniqueId.Segment lastSegment = testDescriptor.getUniqueId().getLastSegment();
+    String type = lastSegment.getType();
+    if ("class".equals(type)) {
+      AtomicReference<Exception> failure = new AtomicReference<>();
+      Thread threadPerClass =
+          new Thread(
+              () -> {
+                try {
+                  testTask.execute();
+                } catch (Exception e) {
+                  failure.set(e);
+                }
+              },
+              "TEST THREAD FOR " + lastSegment.getValue());
```

Comment:
I think it's preferable to use the whole test's `uniqueId` for traceability. Last segments may have the same value for multiple tests under the multi-env engine.

Reply:
@snazy WDYT?

Reply:
I'd prefer to defer that to the JUnit-PR. Also not sure how that would help in real life.